File: weaponOptimizer/jsFiles/buildOptimizer.js
/**
 * Max out the AR of any weapon, given the weapon's level, infusions, starting class, and # of levels to contribute to damage scaling.
 * Boc's Brilliant Build Optimizer
 */
function main() {

//First, get the contants from the spreadsheet for the weapon AR formula. This takes the longest amount of time.
var constants = getWeaponFormulaConstants()
console.log('Done constants');

//On Button press, get user entered information
var numOfLevels = SpreadsheetApp.getActiveSheet().getRange(4, 8).getValue();
var startingClass = SpreadsheetApp.getActiveSheet().getRange(6,8).getValue().toLowerCase();

//Get the wepon requirement levels & scaling of the chosen weapon.
var weaponReq = {'strength':SpreadsheetApp.getActiveSheet().getRange(3,2).getValue(),'dexterity':SpreadsheetApp.getActiveSheet().getRange(3,3).getValue(),'intelligence':SpreadsheetApp.getActiveSheet().getRange(3,4).getValue(),'faith':SpreadsheetApp.getActiveSheet().getRange(3,5).getValue(),'arcane':SpreadsheetApp.getActiveSheet().getRange(3,6).getValue()};

var scaling = [SpreadsheetApp.getActiveSheet().getRange(4,2).getValue(),SpreadsheetApp.getActiveSheet().getRange(4,3).getValue(),SpreadsheetApp.getActiveSheet().getRange(4,4).getValue(),SpreadsheetApp.getActiveSheet().getRange(4,5).getValue(),SpreadsheetApp.getActiveSheet().getRange(4,6).getValue()];
var scalingStats = getScalingStats(scaling);

//Get classes from eldenring.fanapis.com
var classes = JSON.parse(UrlFetchApp.fetch("https://eldenring.fanapis.com/api/classes").getContentText());

//Call optimizeBuildStats() based off of which class the user chose.
for(var i = 0; i < classes['data'].length; i++){
    if(startingClass == classes['data'][i]['name'].toLowerCase()){
    startingStats = classes['data'][i]['stats'];
    optimizeBuildStats(numOfLevels, startingStats, weaponReq, scalingStats, constants);
    }
}
}

////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//Works for weapons that scale off of 0,1, or 2 stats. 3 stats takes too long due to using calculator. 4 or 5 is not supported. (I'm not sure if 4 or 5 is even possible).
function optimizeBuildStats(numOfLevels, startingStats, weaponReq, scalingStats, constants){

//If too many levels are enterted, and all scaling stats can be maxed to 99, return every value being 99.
if(isNumOfLevelsOver(numOfLevels, startingStats, scalingStats)){
    setStats({'strength': 99, 'dexterity': 99, 'intelligence': 99, 'faith': 99, 'arcane': 99, 'ar': null});
    return 1;
}

//Get min stats, then get the number of levels to be initially added to the stats that scale.
var minStats = getMinStats(startingStats, weaponReq);
numOfLevels = getLevelsAfterStatReq(numOfLevels, scalingStats, startingStats, weaponReq)
numOfLevelsStack = getInitialAddedLevelsPerStat(minStats, scalingStats, numOfLevels); 

//Create the starting state
var startState = {'strength': minStats['strength'], 'dexterity': minStats['dexterity'], 'intelligence': minStats['intelligence'], 'faith': minStats['faith'], 'arcane': minStats['arcane'], 'ar': null};

//Add initial levels to stats that scale.
for(var i = 0; i < scalingStats.length; i++){
    startState[scalingStats[i]] += numOfLevelsStack[i];
}

//Get AR of this state
startState['ar'] = getWeaponAR(startState['strength'], startState['dexterity'], startState['intelligence'], startState['faith'], startState['arcane'], constants); 

//If no stats scale or only one stat scales, return startState
if(scalingStats.length == 0 || scalingStats.length == 1)
    return 1;

//Min-max for state with the highest AR (Valid state cannot have a stat under the stat minimum defined in minStats, or total # of levels != starting # of levels)
/** Brute Force approach (although other methods won't be much better) */
var optimialBuild = {...startState}; //Initially the start state
var stateStack = [startState];

//With 2 scalinf stats- branching factor of 1. With 3, branching factor of 2.
//Feel like im thinking of this the wrong way...
var statesVisited = {};
statesVisited[JSON.stringify(startState)] = true;
while(stateStack.length > 0){
    var state = {...stateStack.shift()}
    
    //console.log(state['strength'], state['dexterity'], state['faith']);

    if(state[scalingStats[0]] - 1 >= minStats[scalingStats[0]]){

    state[scalingStats[0]] -= 1;
    for(var i = 1; i < scalingStats.length; i++){
        var newState = {...state};
        if(newState[scalingStats[i]]+1 <= 99){
        newState[scalingStats[i]] += 1;
        newState['ar'] = getWeaponAR(newState['strength'], newState['dexterity'], newState['intelligence'], newState['faith'], newState['arcane'], constants); 

        if(newState['ar'] > optimialBuild['ar']){
            optimialBuild = {...newState};
        }
        
        if(statesVisited[JSON.stringify(newState)] == undefined){
            stateStack.push({...newState});
            statesVisited[JSON.stringify(newState)] = true;
        }
        }
    }
    }
}

setStats(optimialBuild);
console.log("Done Optimization: " + optimialBuild['ar']);
return 1;
}

////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/** Helper functions */
//Get the attack rating from the spreadsheet (Not used anymore)
function getTotalAR(state){
setStats(state)
return SpreadsheetApp.getActiveSheet().getRange(10, 7).getValue();
}

//Set the stats on the spreadsheet to the given stats. {str: ...}
function setStats(stats){
SpreadsheetApp.getActiveSheet().getRange(2, 2).setValue(stats['strength']);
SpreadsheetApp.getActiveSheet().getRange(2, 3).setValue(stats['dexterity']);
SpreadsheetApp.getActiveSheet().getRange(2, 4).setValue(stats['intelligence']);
SpreadsheetApp.getActiveSheet().getRange(2, 5).setValue(stats['faith']);
SpreadsheetApp.getActiveSheet().getRange(2, 6).setValue(stats['arcane']);
}

//Return an array with the names of the stats that scale with the given weapon.
function getScalingStats(scaling){
var scalingStats = []
for(var i = 0; i < scaling.length; i++){
    if(scaling[i]!='-'){
    switch(i){
        case 0:
        scalingStats.push('strength');
        break;
        case 1:
        scalingStats.push('dexterity');
        break;
        case 2:
        scalingStats.push('intelligence');
        break;
        case 3:
        scalingStats.push('faith');
        break;
        case 4:
        scalingStats.push('arcane');
        break;
    }
    }
}
return scalingStats;
}

//Get the minimum stats for each stat line. For each stat this is either the default stat of the starting class, or the weapon required stat, whichever is larger.
function getMinStats(startingStats, weaponReq){


//Minimum stats based off of starting class
var minStr = Math.max(parseInt(startingStats['strength']),weaponReq['strength']);
var minDex = Math.max(parseInt(startingStats['dexterity']),weaponReq['dexterity']);
var minInt = Math.max(parseInt(startingStats['intelligence']),weaponReq['intelligence']);
var minFai = Math.max(parseInt(startingStats['faith']),weaponReq['faith']);
var minArc = Math.max(parseInt(startingStats['arcane']),weaponReq['arcane']);
return {'strength':minStr, 'dexterity':minDex, 'intelligence':minInt, 'faith':minFai, 'arcane':minArc};
}

//Get the amount of levels taken from weapon requirments, that cannot be used in optimizing.
function getLevelsAfterStatReq(numOfLevels, scalingStats, startingStats, weaponReq){
var levelsFromStatReq = 0;

for(var i = 0; i < scalingStats.length; i++){
    if(parseInt(startingStats[scalingStats[i]]) < weaponReq[scalingStats[i]]){
    levelsFromStatReq += weaponReq[scalingStats[i]] - parseInt(startingStats[scalingStats[i]]);
    }
}

return numOfLevels - levelsFromStatReq;
}

//If the user entered enough stats to max out each statline that scales, return true.
function isNumOfLevelsOver(numOfLevels, startingStats, scalingStats){
//Get total number of levels already in scaled stats
var startScalingTotal = 0;
for(var i = 0; i < scalingStats.length; i++){
    startScalingTotal += startingStats[scalingStats[i]]
}

//If the user entered too many levels, then true
if(numOfLevels >= (scalingStats.length * 99)-startScalingTotal){
    return false;
}
else{
    return true;
}
}

//Get the amount of levels to be initally added to each scaling stat line based of the number of levels given, and the min stats of the character. Returns an array of length = # of scaling stats, with each entry being between minStat and 99.
function getInitialAddedLevelsPerStat(minStats, scalingStats, numOfLevels){
var numOfLevelsStack = Array(scalingStats.length).fill(0);
for(var i = 0; i < scalingStats.length; i++){
    if(minStats[scalingStats[i]] + numOfLevels > 99){
    numOfLevelsStack[i] = 99-minStats[scalingStats[i]];
    numOfLevels -= numOfLevelsStack[i];
    }
    else{
    numOfLevelsStack[i] += numOfLevels
    break;
    }
}

return numOfLevelsStack;
}

File: weaponOptimizer/pyFiles/buildOptimizer.py
'''
Build Optimizer for a given weapon.
'''
#Import needed libraries
import requests
import simplejson as JSON
import calcWeaponAR as calcWeapon
import pandas as pd

#Declare global variables, this being dataframes build from CSV files holding weapon information.
df_attack = pd.read_csv(r'eldenring\weaponOptimizer\pyFiles\.csv\Attack.csv')
df_scaling = pd.read_csv(r'eldenring\weaponOptimizer\pyFiles\.csv\Scaling.csv')
df_extraData = pd.read_csv(r'eldenring\weaponOptimizer\pyFiles\.csv\Extra_Data.csv')
df_elementParam = pd.read_csv(r'eldenring\weaponOptimizer\pyFiles\.csv\AttackElementCorrectParam.csv')
df_calcCorrect = pd.read_csv(r'eldenring\weaponOptimizer\pyFiles\.csv\CalcCorrectGraph_ID.csv')

### Main Function ### Takes in the user input, gets the needed information from the CSV files, and calculates the optimal stats needed for the given weapon. User input includes: Weapon Name, Weapon Level, Weapon Affinity(if any), isTwoHanded, Starting class, and the number of levels the user has left in their build to spend on damage stats (str, dex, int, fai, arc).
def main():
    #These var should be entered from user in front-end
    weaponName = "Erdsteel Dagger"
    weaponLevel = str(25)
    isTwoHanded = True
    numOfLevels = 90
    startingClass = "prophet"
    
    #Get if weapon upgrades to +10 or +25, and if it can have affinities.
    affinities = getAffinities(weaponName)
    isSomberWeapon = True if len(affinities) == 1 else False

    #If user enters a weapon level above 10 for a somber weapon, set the weapon level to 10 (Max level)
    weaponLevel = str(10) if (isSomberWeapon and int(weaponLevel) > 10) else weaponLevel

    #Get starting class stats
    startingStats = getStartingStats(startingClass)
    
    #If somber, run once, if can have affinities, run once for each affinity
    for i in range(len(affinities)):
        #If non-somber, then get the next affinity and update weaon name. Else, weapon becomes weaponName.
        if not isSomberWeapon:
            weaponAffinity = affinities[i]
            weapon = weaponAffinity + " " + weaponName
        else:
            weaponAffinity = ""
            weapon = weaponName

        #Get constants for the weapon AR formula for the given weapon. This comes from CSV files parsed by Pandas Dataframes.
        constants = calcWeapon.getWeaponFormulaConstants(weapon,weaponLevel,isTwoHanded)
        print('Done constants')

        #Get the wepon requirement levels & scaling of the chosen weapon. (Repeated work with H2 and H4)
        H2 = int(df_attack.loc[df_attack['Name'] == weapon].index[0])
        H4 = int(df_scaling.columns.get_loc("Str +" + str(weaponLevel)))

        result = df_extraData.iloc[H2,5:10]
        weaponReq = {'strength':int(result[0]),'dexterity':int(result[1]),'intelligence':int(result[2]),'faith':int(result[3]),'arcane':int(result[4])}
        
        scaling = df_scaling.iloc[H2,H4:H4+5].astype(float)
        scalingStats = getScalingStats(scaling)

        #Call optimizeBuildStats()
        print("Affinity: " + weaponAffinity)
        optimizeBuildStats(numOfLevels, startingStats, weaponReq, scalingStats, constants)
                
################################################################################################################
#Works for weapons that scale off of 0,1, or 2 stats. 3 stats takes too long due to using calculator. 4 or 5 is not supported. Need to change algorithm to support 4 & 5.
def optimizeBuildStats(numOfLevels, startingStats, weaponReq, scalingStats, constants):

    #If too many levels are enterted, and all scaling stats can be maxed to 99, return every value being 99.
    if(isNumOfLevelsOver(numOfLevels, startingStats, scalingStats)):
        print("Enough levels to max out every damage stat.")
        return 1
    
    #Get min stats, then get the number of levels to be initially added to the stats that scale.
    minStats = getMinStats(startingStats, weaponReq)
    numOfLevels = getLevelsAfterStatReq(numOfLevels, scalingStats, startingStats, weaponReq)
    numOfLevelsStack = getInitialAddedLevelsPerStat(minStats, scalingStats, numOfLevels)

    #Create the starting state
    startState = {'strength': minStats['strength'], 'dexterity': minStats['dexterity'], 'intelligence': minStats['intelligence'], 'faith': minStats['faith'], 'arcane': minStats['arcane'], 'ar': None}
    statesVisited = {}
    optimialBuild = None

    #Loop through starting here, intililize stateVisted above, each loop has starting stats changed.
    for stat in range(len(scalingStats)):

        state = startState.copy()

        index = 0
        for i in range(stat, len(scalingStats)):
           state[scalingStats[i]] += numOfLevelsStack[index]
           index+=1
        if stat > 0:
            for i in reversed(range(stat-1, 0)):
                state[scalingStats[i]] += numOfLevelsStack[index]
                index+=1
        
        statesVisited[str(state)] = True

        state['ar'] = calcWeapon.getWeaponAR(state['strength'], state['dexterity'], state['intelligence'], state['faith'], state['arcane'], constants)

        #If no stats scale or only one stat scales, return startState
        if len(scalingStats) == 0 or len(scalingStats) == 1:
            return 1

        #Min-max for state with the highest AR (Valid state cannot have a stat under the stat minimum defined in minStats, or total # of levels != starting # of levels)
        #Brute Force approach (although other methods won't be much better) */
        #optimialBuild = state.copy() #Initially the start state
        if optimialBuild is None:
            optimialBuild = state.copy()
        
        #Create a queue to hold the state nodes, starting with the startState. A dict is used to keep track of whether or not a node has been visited, so that the same node is not added back into the queue.
        stateStack = [state]
        
        #Note: With 2 scaling stats, there is a branching factor of 1. With 3 sclaing stats, branching factor of 2.
        #While our queue is not empty, pop the queue.
        while len(stateStack) > 0:
            #Get current state
            state = stateStack.pop().copy()
            
            #Subtract 1 from the first scaling stat
            if(state[scalingStats[stat]] - 1 >= minStats[scalingStats[stat]]):
                state[scalingStats[stat]] -= 1

                #For every other scaling stat, create a state in which the level taken away from the first scaling stat was added to the scaling stat. This (has the potential to) create multiple states.
                for i in range(stat+1, len(scalingStats)):
                    newState = state.copy()
                    #If the stat is not maxed out (less than 99), then add 1 to that stat.
                    if(newState[scalingStats[i]]+1 <= 99):
                        newState[scalingStats[i]] += 1
                        newState['ar'] = calcWeapon.getWeaponAR(newState['strength'], newState['dexterity'], newState['intelligence'], newState['faith'], newState['arcane'], constants);

                        #If the new state has a higher AR then the current optimal state, make the optimal state the new state.
                        if(newState['ar'] > optimialBuild['ar']):
                            optimialBuild = newState.copy()
                        
                        #If the children states from this state have not been expanded, add them tho the queue, and mark this current state as expanded.
                        if(str(newState) not in statesVisited):
                            stateStack.append(newState.copy())
                            statesVisited.update({str(newState):True})
                #From stat - 1 to 0
                for i in reversed(range(stat-1, 0)):
                    newState = state.copy()
                    #If the stat is not maxed out (less than 99), then add 1 to that stat.
                    if(newState[scalingStats[i]]+1 <= 99):
                        newState[scalingStats[i]] += 1
                        newState['ar'] = calcWeapon.getWeaponAR(newState['strength'], newState['dexterity'], newState['intelligence'], newState['faith'], newState['arcane'], constants);

                        #If the new state has a higher AR then the current optimal state, make the optimal state the new state.
                        if(newState['ar'] > optimialBuild['ar']):
                            optimialBuild = newState.copy()
                        
                        #If the children states from this state have not been expanded, add them tho the queue, and mark this current state as expanded.
                        if(str(newState) not in statesVisited):
                            stateStack.append(newState.copy())
                            statesVisited.update({str(newState):True})
                        
    #Done with algorithm, print results to console (for now).
    #print("Done Optimization")
    #print("Str: " + str(optimialBuild['strength']))
    #print("Dex: " + str(optimialBuild['dexterity']))
    #print("Int: " + str(optimialBuild['intelligence']))
    #print("Fai: " + str(optimialBuild['faith']))
    #print("Arc: " + str(optimialBuild['arcane']))
    print("AR: " + str(optimialBuild["ar"]))
   
    return optimialBuild

################################################################################################################
### Helper functions ###
#Return an array with the names of the stats that scale with the given weapon.
def getScalingStats(scaling):
  scalingStats = []

  for i in range(len(scaling)):
    if(scaling[i]!=0):
        if(i == 0):
            scalingStats.append('strength')
        elif(i==1):
            scalingStats.append('dexterity')
        elif(i==2):
            scalingStats.append('intelligence')
        elif(i==3):
            scalingStats.append('faith')
        elif(i==4):
            scalingStats.append('arcane')
        
  return scalingStats

def getMinStats(startingStats, weaponReq):
    #Minimum stats based off of starting class
    minStr = max(int(startingStats['strength']),weaponReq['strength'])
    minDex = max(int(startingStats['dexterity']),weaponReq['dexterity'])
    minInt = max(int(startingStats['intelligence']),weaponReq['intelligence'])
    minFai = max(int(startingStats['faith']),weaponReq['faith'])
    minArc = max(int(startingStats['arcane']),weaponReq['arcane'])
    return {'strength':minStr, 'dexterity':minDex, 'intelligence':minInt, 'faith':minFai, 'arcane':minArc}
    
#Get the amount of levels taken from weapon requirments, that cannot be used in optimizing.
def getLevelsAfterStatReq(numOfLevels, scalingStats, startingStats, weaponReq):
    levelsFromStatReq = 0

    for i in range(len(scalingStats)):
        if(int(startingStats[scalingStats[i]]) < weaponReq[scalingStats[i]]):
            levelsFromStatReq += weaponReq[scalingStats[i]] - int(startingStats[scalingStats[i]])
        
    return numOfLevels - levelsFromStatReq

#If the user entered enough stats to max out each statline that scales, return true.
def isNumOfLevelsOver(numOfLevels, startingStats, scalingStats):
    #Get total number of levels already in scaled stats
    startScalingTotal = 0

    for i in range(len(scalingStats)):
        startScalingTotal = startScalingTotal + int(startingStats[scalingStats[i]])

    #If the user entered too many levels, then true
    if(numOfLevels >= (len(scalingStats) * 99) - startScalingTotal):
        return True
    else:
        return False
    
#Get the amount of levels to be initally added to each scaling stat line based of the number of levels given, and the min stats of the character. Returns an array of length = # of scaling stats, with each entry being between minStat and 99.
def getInitialAddedLevelsPerStat(minStats, scalingStats, numOfLevels):
    numOfLevelsStack = [0] * len(scalingStats)

    for i in range(len(scalingStats)):
        if(minStats[scalingStats[i]] + numOfLevels > 99):
            numOfLevelsStack[i] = 99-minStats[scalingStats[i]]
            numOfLevels -= numOfLevelsStack[i]
        else:
            numOfLevelsStack[i] += numOfLevels
            break
    return numOfLevelsStack
    
#Get the weapon's affinities
def getAffinities(weaponName):
    H2 = int(df_extraData.loc[df_extraData['Name'] == weaponName].index[0])

    return [""] if int(df_extraData.iloc[H2,4]) == 10 else ["Heavy","Keen","Quality","Flame Art","Sacred","Magic","Cold","Fire","Lightning","Poison","Blood","Occult"]

#Get starting stats based off of starting class
def getStartingStats(startingClass):
    classes = JSON.loads(requests.get("https://eldenring.fanapis.com/api/classes").text)
    for i in range(len(classes['data'])):
        if(startingClass == classes['data'][i]['name'].lower()):
            return classes['data'][i]['stats']
                
main()